packages/wonder-blocks-data/util/response-cache.js

```diff
@@ -0,0 +1,149 @@
+        const requestType = handler.type;
+
+        // Ensure we have a cache location for this handler type.
+        this._cache[requestType] = this._cache[requestType] || {};
+
+        // Cache the data.
+        const key = handler.getKey(options);
+        this._cache[requestType][key] = Object.freeze(entry);
+        return this._cache[requestType][key];
+    }
+
+    /**
+     * Initialize the cache from a given cache state.
+     *
+     * This can only be called if the cache is not already in use.
+     */
+    initialize = (source: Cache): void => {
+        if (Object.keys(this._cache).length !== 0) {
+            throw new Error(
+                "Cannot initialize data response cache more than once",
+            );
+        }
+
+        try {
+            /**
+             * Object.assign only performs a shallow clone.
+             * So we deep clone it and then assign the clone values to our
+             * internal cache.
+             */
+            const cloneInitCache = deepClone(source);
+            Object.assign(this._cache, cloneInitCache);
+        } catch (e) {
+            throw new Error(
+                `An error occurred trying to initialize the data response cache: ${e}`,
+            );
+        }
+    };
+
+    /**
+     * Cache data for a specific response.
+     */
+    cacheData = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+        data: TData,
+    ): CacheEntry<TData> => {
+        return this._setCacheEntry(handler, options, {data});
+    };
+
+    /**
+     * Cache an error for a specific response.
+     */
+    cacheError = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+        error: Error | string,
+    ): CacheEntry<TData> => {
+        const errorMessage = typeof error === "string" ? error : error.message;
+        return this._setCacheEntry(handler, options, {error: errorMessage});
+    };
+
+    /**
+     * Retrieve data from our cache.
+     */
+    getEntry = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+    ): ?$ReadOnly<CacheEntry<TData>> => {
+        const requestType = handler.type;
+
+        // Get the subcache for the handler.
+        const handlerCache = this._cache[requestType];
+        if (!handlerCache) {
+            return null;
+        }
+
+        const key = handler.getKey(options);
+        if (handler.invalidateCache(options)) {
+            delete handlerCache[key];
+        }
+
+        // Get the response.
+        const entry = handlerCache[key];
+        return entry == null ? null : entry;
+    };
+
+    /**
+     * Deep clone the cache.
+     */
+    clone = (): $ReadOnly<Cache> => {
+        try {
+            return deepClone(this._cache);
+        } catch (e) {
+            throw new Error(
+                `An error occurred while trying to clone the cache: ${e}`,
+            );
+        }
+    };
```

Comment:
Why don't we wrap the `_cache` in a `ResponseCache` before returning it?

Reply:
It is a `ResponseCache`. It's just that we can't have both the `ResponseCache` type and the `ResponseCache` class co-exist with the same name.

We don't need this to be a `ResponseCache` class because that's not the purpose of this method.

Reply:
It's a little unexpected that's all.  Usually a `clone()` method on a class would return a new instance of the class with a copy of the data from the original.  That isn't what's happening in this case.  It's fine if that's what we want, but I do think it's worth making note of it in the docstring.